website/blog/buck2.md

```diff
@@ -563,3 +566,116 @@ I hope the current excitement could grow a community and ecosystem around Buck2
+-        tools = [
++        srcs = [
+             ":x_{}".format(n),
+             ":y_{}".format(n),
+         ],
++        outs = ["z_{}.txt".format(n)],
++        cmd = SLEEP_CMD + "cat $(location :x_{n}) $(location :y_{n}) > $@".format(n = n),
+     )
+     for n in SIZE_RANGE
+ ]
+
+ genrule(
+     name = "xyz",
++    srcs = [
++        ":z_{}".format(n)
++        for n in SIZE_RANGE
++    ],
+     outs = ["xyz.txt"],
+     # cat $(location :z_0) $(location :z_1) ... $(location :z_999) > $@,
+     cmd = SLEEP_CMD + "cat " + " ".join([
+         "$(location :z_{})".format(n)
+         for n in SIZE_RANGE
+     ]) + " > $@",
+-    tools = [
+-        ":z_{}".format(n)
+-        for n in SIZE_RANGE
+-    ],
+ )
+```
+
+Using `tools` would trigger Bazel to configure twice the number of actions underneath due to Bazel's special treament for tool dependencies.
+
+And fewer actions means faster build.
+For example, here is the new benchmark when build and test everything with Bazel.
+
+```bash
+> hyperfine --prepare 'bazel clean' --warmup 1 'bazel test --config=local //...'
+Benchmark 1: bazel test --config=local //...
+  Time (mean ± σ):     42.580 s ±  1.881 s    [User: 0.039 s, System: 0.048 s]
+  Range (min … max):   38.127 s … 45.369 s    10 runs
+```
+
+So I reran the benchmark and here are the new timings.
+
+| Tools       |          Local Build | RBE (with Remote Cache) | RBE (without Remote Cache) |
+| :---------- | -------------------: | ----------------------: | -------------------------: |
+| buck2 (old) | ` 1.175 s ± 2.183 s` |    ` 7.937 s ± 0.487 s` |     `203.431 s ± 29.882 s` |
+| bazel (old) | `65.631 s ± 0.460 s` |    ` 9.798 s ± 0.515 s` |    `106.062 s ± ‎ 3.106 s` |
+| bazel (new) | `42.580 s ± 1.881 s` |    `10.047 s ± 0.517 s` |    ` 73.009 s ± ‎ 2.975 s` |
+
+This means that if setup correctly, Bazel performance is quite competitive with Buck2.
+Differences in a few seconds could be negligible for most use cases.
+
+Tobias also noted that my current Bazel configuration is uploading Build Events, which may incur additional network latency.
+This could be further be improved by moving the Build Event upload to a background task and defer build artifact downloads.
+
+Here is a quick comparision with RBE and Remote Cache on different modes:
+
+```bash
+> hyperfine --prepare 'bazel clean' \
+            --warmup 1 \
+            'bazel test --config=remote //...' \
+            'bazel test --config=remote //... --bes_upload_mode=fully_async' \
+            'bazel test --config=remote //... --bes_upload_mode=fully_async --remote_download_toplevel' \
+            'bazel test --config=remote //... --bes_upload_mode=fully_async --remote_download_minimal'
+
+Benchmark 1: bazel test --config=remote //...
+  Time (mean ± σ):      9.460 s ±  0.540 s    [User: 0.023 s, System: 0.031 s]
+  Range (min … max):    8.473 s … 10.107 s    10 runs
+
+Benchmark 2: bazel test --config=remote //... --bes_upload_mode=fully_async
```

Comment:
Is the buck2 build configured to use its BES equivalent? If not, I'm not sure it makes sense to compare against Bazel with BES enabled

Reply:
Updated. PTAL 🙏